apps/derive/elpi/lens_laws.elpi

```diff
@@ -0,0 +1,89 @@
+/* Equations on lenses                                                       */
+/* license: GNU Lesser General Public License Version 2.1 or later           */
+/* ------------------------------------------------------------------------- */
+
+namespace derive.lens-laws {
+
+pred declare-law1 i:prop.
+declare-law1 (lens-db I C) :- std.do! [
+  coq.env.indt-decl I Decl,
+  law1 Decl (global (const C)) (global (indt I)) Bo Ty,
+  Name is {coq.gref->id (const C)} ^ "_law1",
+  coq.env.add-const Name Bo Ty @opaque! _,
+].
+pred law1 i:indt-decl, i:term, i:term, o:term, o:term.
+law1 (parameter ID _ Ty Rest) Lens Ind {{ fun p : lp:Ty => lp:(Bo p) }} {{ forall p : lp:Ty, lp:(BoTy p) }} :- !,
+  @pi-parameter ID Ty p\ law1 (Rest p) {coq.mk-app Lens [p]} {coq.mk-app Ind [p]} (Bo p) (BoTy p).
+law1 _ Lens IT Bo Ty :-
+  BoS = {{ fun r => lp:{{ {coq.build-match {{r}} IT (law1-rty.aux Lens) law1-bo.aux } }} }},
+  TyS = {{ lib:elpi.derive.lens.law1 lp:Lens }},
+  std.assert-ok! (coq.elaborate-ty-skeleton TyS _ Ty) "ty",
+  std.assert-ok! (coq.elaborate-skeleton BoS Ty Bo) "bo".
+law1-rty.aux L _ Vs _ {{ lib:elpi.derive.lens.law1_on lp:L lp:R }} :- std.last Vs R.
+law1-bo.aux _ _ _ _ {{ fun x => lib:@elpi.erefl _ _ }}.
+
+pred declare-law2 i:prop.
+declare-law2 (lens-db I C) :- std.do! [
+  coq.env.indt-decl I Decl,
+  law2 Decl (global (const C)) (global (indt I)) Bo Ty,
+  Name is {coq.gref->id (const C)} ^ "_law2",
+  coq.env.add-const Name Bo Ty @opaque! _,
+].
+pred law2 i:indt-decl, i:term, i:term, o:term, o:term.
+law2 (parameter ID _ Ty Rest) Lens Ind {{ fun p : lp:Ty => lp:(Bo p) }} {{ forall p : lp:Ty, lp:(BoTy p) }} :- !,
+  @pi-parameter ID Ty p\ law2 (Rest p) {coq.mk-app Lens [p]} {coq.mk-app Ind [p]} (Bo p) (BoTy p).
+law2 _ Lens IT Bo Ty :-
+  BoS = {{ fun r => lp:{{ {coq.build-match {{r}} IT (law2-rty.aux Lens) law2-bo.aux } }} }},
+  TyS = {{ lib:elpi.derive.lens.law2 lp:Lens }},
+  std.assert-ok! (coq.elaborate-ty-skeleton TyS _ Ty) "ty",
+  std.assert-ok! (coq.elaborate-skeleton BoS Ty Bo) "bo".
```

Comment:
Is `coq.elaborate-skeleton` taking something like a term with holes and passing it through the elaborator?

Reply:
yes, I could not find a better name for this kind of terms. Also, it does not modify it in place (it does not assign the holes), it uses it as a "skeleton".

Reply:
I did not have the time to document this file, I tried to do a better job in the other one (the one synthesizing lenses)